File: rust/crates/harness/src/projection.rs
//! Explicit, provenance-preserving selection from canonical history to model context.

pub use crate::conversation::ModelContextSelection;
use crate::{
    Error, Result,
    conversation::{
        Attachment, ContentResidencyVerifier, ConversationMessage, ConversationState, FileRef,
        MessageKind, ReferencedAttachments,
    },
    model::{FileProjectionPolicy, ModelContent, ModelContentPart, ModelMessage, ModelRole},
    tool::ToolInvocation,
};
use futures::future::BoxFuture;
use serde::{Deserialize, Serialize, de::DeserializeOwned};
use std::collections::HashMap;
use uuid::Uuid;

/// Provider-owned, grant-checked resolver for a complete attachment manifest.
pub trait AttachmentListResolver: Send + Sync {
    /// Verifies the pinned manifest bytes and returns its complete ordered list.
    fn resolve<'a>(
        &'a self,
        manifest: &'a FileRef,
        item_count: u32,
    ) -> BoxFuture<'a, Result<Vec<Attachment>>>;

    /// Reads one bounded, exact structured artifact for tool-context projection.
    fn read<'a>(&'a self, _file: &'a FileRef) -> BoxFuture<'a, Result<Vec<u8>>> {
        Box::pin(async {
            Err(Error::Unsupported(
                "tool artifact resolution is unavailable".into(),
            ))
        })
    }
}

impl<T: ContentResidencyVerifier + ?Sized> AttachmentListResolver for T {
    fn resolve<'a>(
        &'a self,
        manifest: &'a FileRef,
        item_count: u32,
    ) -> BoxFuture<'a, Result<Vec<Attachment>>> {
        Box::pin(async move { self.load_manifest(manifest, item_count).await })
    }

    fn read<'a>(&'a self, file: &'a FileRef) -> BoxFuture<'a, Result<Vec<u8>>> {
        ContentResidencyVerifier::read(self, file)
    }
}

/// Transient model context with explicit canonical provenance.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct SelectedModelContext {
    /// Selection and exact history revision.
    pub selection: ModelContextSelection,
    /// Typed model-visible values; do not append them to conversation history.
    pub messages: Vec<ModelMessage>,
}

impl SelectedModelContext {
    /// Checks that a projected selection preserves one-to-one provenance and
    /// ends in the exact current user input rather than duplicating it.
    pub fn validate_for_input(&self, input: &ModelContent) -> Result<()> {
        if self.messages.is_empty()
            || self.messages.len() != self.selection.message_ids.len()
            || self
                .messages
                .last()
                .is_none_or(|last| last.role != ModelRole::User || &last.content != input)
        {
            return Err(Error::Invalid(
                "selected context must end with the current user input".into(),
            ));
        }
        Ok(())
    }
}

/// Selects a bounded ordered subset. Text and attachments stay as refs;
/// structured tool artifacts are resolved only within the render bound.
pub async fn select_model_context<R: AttachmentListResolver + ?Sized>(
    conversation: &ConversationState,
    selection: ModelContextSelection,
    resolver: &R,
    maximum_messages: usize,
    maximum_attachments: usize,
    maximum_render_bytes: u64,
) -> Result<SelectedModelContext> {
    selection.validate(conversation)?;
    if maximum_messages == 0
        || maximum_render_bytes == 0
        || selection.message_ids.len() > maximum_messages
    {
        return Err(Error::Invalid(
            "selected message count exceeds projection limit".into(),
        ));
    }
    let by_id = conversation
        .messages
        .iter()
        .map(|message| (message.id, message))
        .collect::<HashMap<_, _>>();
    if by_id.len() != conversation.messages.len() {
        return Err(Error::Invalid(
            "conversation has duplicate message identities".into(),
        ));
    }
    let mut messages = Vec::with_capacity(selection.message_ids.len());
    let mut tool_calls: HashMap<Uuid, (String, String)> = HashMap::new();
    for id in &selection.message_ids {
        let message = by_id
            .get(id)
            .copied()
            .ok_or_else(|| Error::Invalid("selected conversation message is missing".into()))?;
        message.validate()?;
        if message.kind == MessageKind::ToolCall {
            let invocation: ToolInvocation =
                read_json_artifact(resolver, &message.content, maximum_render_bytes).await?;
            invocation.validate()?;
            if message.tool_call_id.as_deref() != Some(invocation.call_id.as_str()) {
                return Err(Error::Invalid(
                    "tool call artifact identity does not match its record".into(),
                ));
            }
            tool_calls.insert(
                message.id,
                (invocation.call_id.clone(), invocation.name.clone()),
            );
            messages.push(ModelMessage {
                role: ModelRole::Assistant,
                content: ModelContent::Part(ModelContentPart::ToolCall {
                    call_id: invocation.call_id,
                    name: invocation.name,
                    arguments: invocation.arguments,
                }),
            });
            continue;
        }
        if message.kind == MessageKind::ToolResult {
            // The complete result remains an immutable artifact. Only its
            // separately bounded projection is model-visible here.
            if message.content.descriptor().media_type() != "application/json" {
                return Err(Error::Invalid(
                    "tool result artifact type is invalid".into(),
                ));
            }
            let call_id = message
                .tool_call_id
                .as_deref()
                .ok_or_else(|| Error::Invalid("tool result lacks call identity".into()))?;
            let (linked_call_id, name) = message
                .reply_to
                .and_then(|id| tool_calls.get(&id))
                .ok_or_else(|| Error::Invalid("selected tool result lacks its call".into()))?;
            if linked_call_id != call_id {
                return Err(Error::Invalid(
                    "selected tool result has a mismatched call".into(),
                ));
            }
            let attachments = resolve_attachments(message, resolver, maximum_attachments).await?;
            let projection = attachments
                .iter()
                .find(|attachment| attachment.label.as_deref() == Some("model_projection"))
                .ok_or_else(|| Error::Invalid("tool result lacks its model projection".into()))?;
            let value =
                read_json_artifact(resolver, &projection.file, maximum_render_bytes).await?;
            messages.push(ModelMessage {
                role: ModelRole::Tool,
                content: ModelContent::Part(ModelContentPart::ToolResult {
                    call_id: call_id.to_owned(),
                    name: name.clone(),
                    value,
                }),
            });
            continue;
        }
        let role = match message.kind {
            MessageKind::System => ModelRole::System,
            MessageKind::User => ModelRole::User,
            MessageKind::Assistant => ModelRole::Assistant,
            _ => {
                return Err(Error::Unsupported(format!(
                    "message kind {:?} requires a specialized model projection",
                    message.kind
                )));
            }
        };
        let primary_policy = match message.content.descriptor().media_type() {
            "image/png" | "image/jpeg" | "image/gif" | "image/webp" => FileProjectionPolicy::Native,
            kind if kind.starts_with("text/")
                && message.content.descriptor().byte_length() <= maximum_render_bytes =>
            {
                FileProjectionPolicy::BoundedFull
            }
            _ => FileProjectionPolicy::Reference,
        };
        let mut parts = vec![ModelContentPart::File {
            file: message.content.clone(),
            policy: primary_policy,
        }];
        if let ReferencedAttachments::Manifest { item_count, .. } = &message.attachments
            && (*item_count as usize) > maximum_attachments
        {
            return Err(Error::Invalid(
                "selected message exceeds attachment projection limit".into(),
            ));
        }
        let attachments = resolve_attachments(message, resolver, maximum_attachments).await?;
        let projected_count = attachments.len().min(1_022);
        let omitted_count = attachments.len() - projected_count;
        for attachment in attachments.into_iter().take(projected_count) {
            attachment.validate()?;
            let policy = match attachment.file.descriptor().media_type() {
                "image/png" | "image/jpeg" | "image/gif" | "image/webp" => {
                    FileProjectionPolicy::Native
                }
                _ => FileProjectionPolicy::Reference,
            };
            parts.push(ModelContentPart::File {
                file: attachment.file,
                policy,
            });
        }
        if omitted_count > 0 {
            parts.push(ModelContentPart::Text {
                text: format!("[{omitted_count} additional attachments omitted from this bounded model context]"),
            });
        }
        messages.push(ModelMessage {
            role,
            content: ModelContent::Parts(parts),
        });
    }
    Ok(SelectedModelContext {
        selection,
        messages,
    })
}

async fn read_json_artifact<T: DeserializeOwned, R: AttachmentListResolver + ?Sized>(
    resolver: &R,
    file: &FileRef,
    maximum_bytes: u64,
) -> Result<T> {
    if file.descriptor().media_type() != "application/json"
        || file.descriptor().byte_length() > maximum_bytes
    {
        return Err(Error::Invalid(
            "tool artifact type or rendering limit is invalid".into(),
        ));
    }
    let bytes = resolver.read(file).await?;
    if bytes.len() as u64 > maximum_bytes {
        return Err(Error::Invalid(
            "resolved tool artifact exceeds rendering limit".into(),
        ));
    }
    file.descriptor().verify(&bytes)?;
    serde_json::from_slice(&bytes)
        .map_err(|error| Error::Invalid(format!("tool artifact is invalid: {error}")))
}

async fn resolve_attachments<R: AttachmentListResolver + ?Sized>(
    message: &ConversationMessage,
    resolver: &R,
    maximum_attachments: usize,
) -> Result<Vec<Attachment>> {
    let attachments = match &message.attachments {
        ReferencedAttachments::Inline { items } => items.clone(),
        ReferencedAttachments::Manifest {
            manifest,
            item_count,
        } => {
            if *item_count as usize > maximum_attachments {
                return Err(Error::Invalid(
                    "selected message exceeds attachment projection limit".into(),
                ));
            }
            let items = resolver.resolve(manifest, *item_count).await?;
            if items.len() != *item_count as usize {
                return Err(Error::Invalid(
                    "attachment manifest count does not match its declared count".into(),
                ));
            }
            items
        }
    };
    if attachments.len() > maximum_attachments {
        return Err(Error::Invalid(
            "selected message exceeds attachment projection limit".into(),
        ));
    }
    Ok(attachments)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::conversation::{
        ConversationMessage, FileDescriptor, VolumeClass, VolumeOwner, VolumeRef,
    };
    use crate::{AgentId, resources::ProviderRef};
    use std::{
        collections::BTreeMap,
        sync::atomic::{AtomicUsize, Ordering},
    };
    use uuid::Uuid;

    struct Resolver {
        items: Vec<Attachment>,
        calls: AtomicUsize,
    }

    impl AttachmentListResolver for Resolver {
        fn resolve<'a>(
            &'a self,
            _manifest: &'a FileRef,
            _item_count: u32,
        ) -> BoxFuture<'a, Result<Vec<Attachment>>> {
            self.calls.fetch_add(1, Ordering::SeqCst);
            Box::pin(async { Ok(self.items.clone()) })
        }
    }

    fn file(agent: AgentId, path: &str, media_type: &str) -> Result<FileRef> {
        FileRef::new(
            VolumeRef::new(
                ProviderRef::new("test", "filesystem", "2")?,
                "private",
                VolumeClass::AgentPrivate,
                VolumeOwner::Agent(agent),
            )?,
            path,
            "v1",
            FileDescriptor::from_bytes(b"data", media_type)?,
            "data",
        )
    }

    #[tokio::test]
    async fn manifest_projection_checks_limit_before_resolution_and_exact_count_after() -> Result<()>
    {
        let agent = AgentId::new();
        let id = Uuid::new_v4();
        let mut conversation = ConversationState::default();
        conversation.bind(agent)?;
        conversation.append(ConversationMessage {
            id,
            sequence: 1,
            kind: MessageKind::User,
            content: file(agent, "message.txt", "text/plain")?,
            attachments: ReferencedAttachments::Manifest {
                manifest: file(
                    agent,
                    "manifest.json",
                    "application/vnd.acyclic.harness.attachments+json",
                )?,
                item_count: 2,
            },
            reply_to: None,
            tool_call_id: None,
            extensions: BTreeMap::new(),
        })?;
        let selection = ModelContextSelection {
            conversation_revision: 1,
            message_ids: vec![id],
        };
        let attachment = Attachment {
            file: file(agent, "image.png", "image/png")?,
            label: None,
        };
        let resolver = Resolver {
            items: vec![attachment.clone()],
            calls: AtomicUsize::new(0),
        };
        assert!(
            select_model_context(
                &conversation,
                selection.clone(),
                &resolver,
                256,
                1,
                128 * 1024
            )
            .await
            .is_err()
        );
        assert_eq!(resolver.calls.load(Ordering::SeqCst), 0);
        assert!(
            select_model_context(
                &conversation,
                selection.clone(),
                &resolver,
                256,
                2,
                128 * 1024
            )
            .await
            .is_err()
        );
        assert_eq!(resolver.calls.load(Ordering::SeqCst), 1);
        let resolver = Resolver {
            items: vec![attachment.clone(), attachment],
            calls: AtomicUsize::new(0),
        };
        let projected =
            select_model_context(&conversation, selection, &resolver, 256, 2, 128 * 1024).await?;
        assert_eq!(projected.messages.len(), 1);
        assert_eq!(resolver.calls.load(Ordering::SeqCst), 1);
        Ok(())
    }
}

File: typescript/packages/harness/src/projection.ts
/** Explicit, ref-only selection from canonical conversation history into model context. */
import { DEFAULT_LIMITS, decodeAttachmentManifest,
  verifyFileBytes, type Attachment, type ConversationMessageId, type ConversationState, type FileRef, type Limits } from "./conversation.js";
import { validateToolName, type ModelContentPart, type ModelMessage } from "./model.js";
import { NativeContracts } from "./native-contracts.js";
import type { ContentBindings } from "./runtime.js";

export interface ModelContextSelection {
  /** Exact conversation tail observed when this selection was made. */
  readonly conversationRevision: bigint;
  /** Ordered subset of canonical message IDs. */
  readonly messageIds: readonly ConversationMessageId[];
}

/** A provider-proven view may contain less than the complete durable history. */
export interface ProjectableConversation extends ConversationState {
  readonly revision: bigint;
}

export interface SelectedModelContext {
  readonly selection: ModelContextSelection;
  /** Volatile provider-neutral projection; never append this to conversation history. */
  readonly messages: readonly ModelMessage[];
}

export interface ConversationProjectionOptions {
  /** Owner-mediated read; possession of a manifest ref is not authorization. */
  readonly resolveManifest?: (file: FileRef) => Promise<Uint8Array>;
  readonly maxManifestBytes?: number;
  readonly maxAttachments?: number;
  readonly maxMessages?: number;
  /** Bound the actual model request even when canonical history has a large referenced list. */
  readonly maxProjectedAttachments?: number;
  /** Owner-mediated exact reads for structured tool artifacts. */
  readonly resolveFile?: (file: FileRef) => Promise<Uint8Array>;
  readonly maxRenderBytes?: number;
  /** Optional owner decoder; its result must match Rust's canonical decoding exactly. */
  readonly decodeManifest?: (manifest: FileRef, bytes: Uint8Array, count: number) => readonly Attachment[] | Promise<readonly Attachment[]>;
  /** Optional additional owning-provider validation; never bypasses Rust. */
  readonly validateMessage?: (message: ConversationState["messages"][number]) => ConversationState["messages"][number] | Promise<ConversationState["messages"][number]>;
}

export interface FileProjectionOptions {
  /** Must enforce the owner's read grant before returning exact bytes. */
  readonly resolveFile?: (file: FileRef) => Promise<Uint8Array>;
  /** Optional additional owner verification; Rust always verifies the descriptor. */
  readonly verifyFile?: (file: FileRef, bytes: Uint8Array) => void | Promise<void>;
  readonly maxResolvedBytes?: number;
}

/** Adapts an exact-volume content binding into a verified model resolver. */
export function verifiedContentResolver(
  content: Pick<ContentBindings, "validate" | "verify" | "read">, limits: Limits,
): (file: FileRef) => Promise<Uint8Array> {
  const validate = content.validate.bind(content);
  const verify = content.verify.bind(content);
  const read = content.read.bind(content);
  return async file => {
    const reference = (await NativeContracts.create()).validate("file_ref", file);
    validate(reference, limits);
    const bytes = await read(reference);
    if (!(bytes instanceof Uint8Array)) throw new TypeError("content provider returned invalid bytes");
    verify(reference, bytes);
    await verifyFileBytes(reference, bytes);
    return Uint8Array.from(bytes);
  };
}

export type ProjectedFile =
  | Readonly<{ kind: "reference"; text: string }>
  | Readonly<{ kind: "text"; text: string }>
  | Readonly<{ kind: "image"; mediaType: string; bytes: Uint8Array }>;

/** Shared fail-closed file boundary for default model adapters. */
export async function projectModelFile(
  part: Extract<ModelContentPart, { kind: "file" }>, options: FileProjectionOptions,
): Promise<ProjectedFile> {
  const file = (await NativeContracts.create()).validate("file_ref", part.file);
  if (part.policy === "reference") {
    const hash = file.descriptor.sha256.map(byte => byte.toString(16).padStart(2, "0")).join("");
    return { kind: "reference", text: `[file: ${file.display_name}; ${file.descriptor.media_type}; sha256=${hash}; bytes=${file.descriptor.byte_length}]` };
  }
  if (part.policy !== "native" && part.policy !== "bounded_full") {
    throw new TypeError("unsupported file projection policy");
  }
  const nativeImage = part.policy === "native"
    && ["image/png", "image/jpeg", "image/gif", "image/webp"].includes(file.descriptor.media_type);
  const boundedText = part.policy === "bounded_full" && file.descriptor.media_type.startsWith("text/");
  if (!nativeImage && !boundedText) {
    throw new TypeError("unsupported file content for selected projection policy");
  }
  const limit = options.maxResolvedBytes ?? 1_048_576;
  if (!Number.isSafeInteger(limit) || limit < 0) throw new TypeError("maxResolvedBytes must be a non-negative safe integer");
  if (options.resolveFile === undefined) throw new TypeError("file byte resolution is unavailable");
  if (file.descriptor.byte_length > limit) throw new TypeError("file exceeds projection byte limit");
  const bytes = await options.resolveFile(file);
  if (!(bytes instanceof Uint8Array) || bytes.byteLength > limit) {
    throw new TypeError("resolved file exceeds projection byte limit");
  }
  await options.verifyFile?.(file, bytes);
  await verifyFileBytes(file, bytes);
  if (nativeImage) {
    return { kind: "image", mediaType: file.descriptor.media_type, bytes: Uint8Array.from(bytes) };
  }
  if (boundedText) {
    return { kind: "text", text: new TextDecoder("utf-8", { fatal: true }).decode(bytes) };
  }
  throw new TypeError("unsupported file content for selected projection policy");
}

export async function selectModelContext(
  conversation: ProjectableConversation,
  selection: ModelContextSelection,
  options: ConversationProjectionOptions,
): Promise<SelectedModelContext> {
  if (typeof conversation.revision !== "bigint" || conversation.revision < 0n
    || typeof selection.conversationRevision !== "bigint"
    || selection.conversationRevision !== conversation.revision) {
    throw new TypeError("model context selection has a stale conversation revision");
  }
  const maxManifestBytes = options.maxManifestBytes ?? 1_048_576;
  const maxAttachments = options.maxAttachments ?? 256;
  const maxMessages = options.maxMessages ?? 256;
  const maxProjectedAttachments = options.maxProjectedAttachments ?? 1_022;
  const maxRenderBytes = options.maxRenderBytes ?? 128 * 1024;
  if (!Number.isSafeInteger(maxManifestBytes) || maxManifestBytes < 0
    || !Number.isSafeInteger(maxAttachments) || maxAttachments < 0
    || !Number.isSafeInteger(maxMessages) || maxMessages <= 0
    || !Number.isSafeInteger(maxProjectedAttachments) || maxProjectedAttachments < 0 || maxProjectedAttachments > 1_022
    || !Number.isSafeInteger(maxRenderBytes) || maxRenderBytes <= 0) {
    throw new TypeError("invalid projection limits");
  }
  if (selection.messageIds.length > maxMessages) throw new TypeError("selected message count exceeds projection limit");
  const byId = new Map(conversation.messages.map(message => [message.id, message]));
  if (byId.size !== conversation.messages.length) throw new TypeError("conversation has duplicate message identities");
  const selected: ModelMessage[] = [];
  const toolCalls = new Map<ConversationMessageId, { callId: string; name: string }>();
  const contracts = await NativeContracts.create();
  let previousSequence = 0n;
  for (const id of selection.messageIds) {
    const raw = byId.get(id);
    if (raw === undefined) throw new TypeError("selected conversation message is missing");
    const canonical = contracts.validate("conversation_message", raw, DEFAULT_LIMITS);
    const message = options.validateMessage === undefined ? canonical
      : contracts.validate("conversation_message", await options.validateMessage(canonical), DEFAULT_LIMITS);
    if (message.sequence <= previousSequence || message.sequence > conversation.revision) {
      throw new TypeError("model context selection is not ordered and unique");
    }
    previousSequence = message.sequence;
    if (message.kind === "tool_call") {
      const invocation = await readJsonArtifact(message.content, options.resolveFile, maxRenderBytes);
      if (!isRecord(invocation) || typeof invocation.call_id !== "string"
        || typeof invocation.name !== "string" || invocation.call_id !== message.tool_call_id
        || !invocation.name) {
        throw new TypeError("tool call artifact identity does not match its record");
      }
      validateToolName(invocation.name);
      toolCalls.set(message.id, { callId: invocation.call_id, name: invocation.name });
      selected.push({ role: "assistant", content: { kind: "tool_call", callId: invocation.call_id,
        name: invocation.name, arguments: invocation.arguments } });
      continue;
    }
    if (message.kind === "tool_result") {
      // The complete output remains an artifact; the model sees only the
      // separately bounded projection attachment.
      if (message.content.descriptor.media_type !== "application/json") {
        throw new TypeError("tool result artifact type is invalid");
      }
      const callId = message.tool_call_id;
      const linked = message.reply_to === null ? undefined : toolCalls.get(message.reply_to);
      if (callId === null || linked === undefined || linked.callId !== callId) {
        throw new TypeError("selected tool result lacks its call");
      }
      const artifacts = await resolveAttachments(message.attachments, options.resolveManifest ?? options.resolveFile,
        maxManifestBytes, maxAttachments, options.decodeManifest);
      const projection = artifacts.find(artifact => artifact.label === "model_projection");
      if (projection === undefined) throw new TypeError("tool result lacks its model projection");
      const value = await readJsonArtifact(projection.file, options.resolveFile, maxRenderBytes);
      selected.push({ role: "tool", content: { kind: "tool_result", callId, name: linked.name, value } });
      continue;
    }
    if (message.kind !== "system" && message.kind !== "user" && message.kind !== "assistant") {
      throw new TypeError(`message kind ${message.kind} requires a specialized model projection`);
    }
    const primaryType = message.content.descriptor.media_type;
    const parts: ModelContentPart[] = [{
      kind: "file", file: message.content,
      policy: ["image/png", "image/jpeg", "image/gif", "image/webp"].includes(primaryType) ? "native"
        : primaryType.startsWith("text/") && message.content.descriptor.byte_length <= maxRenderBytes
          ? "bounded_full" : "reference",
    }];
    const attachments = await resolveAttachments(message.attachments, options.resolveManifest ?? options.resolveFile,
      maxManifestBytes, maxAttachments, options.decodeManifest);
    if (attachments.length > maxAttachments) throw new TypeError("selected message exceeds attachment projection limit");
    for (const attachment of attachments.slice(0, maxProjectedAttachments)) {
      const type = attachment.file.descriptor.media_type;
      parts.push({ kind: "file", file: attachment.file, policy: ["image/png", "image/jpeg", "image/gif", "image/webp"].includes(type) ? "native" : "reference" });
    }
    if (attachments.length > maxProjectedAttachments) {
      parts.push({ kind: "text", text: `[${attachments.length - maxProjectedAttachments} additional attachments omitted from this bounded model context]` });
    }
    selected.push({ role: message.kind, content: parts });
  }
  return Object.freeze({
    selection: Object.freeze({ conversationRevision: selection.conversationRevision, messageIds: Object.freeze([...selection.messageIds]) }),
    messages: Object.freeze(selected),
  });
}

async function readJsonArtifact(file: FileRef,
  resolver: ConversationProjectionOptions["resolveFile"], maxBytes: number): Promise<unknown> {
  const reference = (await NativeContracts.create()).validate("file_ref", file);
  if (reference.descriptor.media_type !== "application/json"
    || reference.descriptor.byte_length > maxBytes) {
    throw new TypeError("tool artifact type or rendering limit is invalid");
  }
  if (resolver === undefined) throw new TypeError("tool artifact resolution is unavailable");
  const bytes = await resolver(reference);
  if (!(bytes instanceof Uint8Array) || bytes.byteLength > maxBytes) {
    throw new TypeError("resolved tool artifact exceeds rendering limit");
  }
  await verifyFileBytes(reference, bytes);
  return (await NativeContracts.create()).decodeModelJson(bytes);
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

async function resolveAttachments(
  value: ConversationState["messages"][number]["attachments"],
  resolver: ((file: FileRef) => Promise<Uint8Array>) | undefined,
  maxBytes: number,
  maxAttachments: number,
  decode: ConversationProjectionOptions["decodeManifest"],
): Promise<readonly Attachment[]> {
  const validated = (await NativeContracts.create()).validate("attachments", value);
  if (validated.kind === "inline") return validated.items;
  if (validated.item_count > maxAttachments) throw new TypeError("selected message exceeds attachment projection limit");
  if (validated.manifest.descriptor.byte_length > maxBytes) throw new TypeError("attachment manifest exceeds projection limit");
  if (resolver === undefined) throw new TypeError("attachment manifest resolution is unavailable");
  const bytes = await resolver(validated.manifest);
  if (!(bytes instanceof Uint8Array) || bytes.byteLength > maxBytes) {
    throw new TypeError("resolved attachment manifest exceeds projection limit");
  }
  const canonical = await decodeAttachmentManifest(validated.manifest, bytes, validated.item_count);
  if (decode !== undefined) {
    const ownerDecoded = await decode(validated.manifest, bytes, validated.item_count);
    const contracts = await NativeContracts.create();
    const canonicalBytes = contracts.encodeAttachmentManifest(canonical);
    const ownerBytes = contracts.encodeAttachmentManifest(ownerDecoded);
    if (canonicalBytes.length !== ownerBytes.length
      || canonicalBytes.some((byte, index) => byte !== ownerBytes[index])) {
      throw new TypeError("owner attachment decoder disagrees with canonical manifest");
    }
  }
  return canonical;
}
